cgen error: undefined symbol 'main__encode_array_T_Array_int'

### Describe the bug

tcc: error: undefined symbol 'main__encode_array_T_Array_int'

### Expected Behavior

```bash
[encode_array.v:7] 'is int': is int
```

### Current Behavior

```bash
==================
/tmp/v_1000/encode_array.10225179450966957427.tmp.c:12239: warning: implicit declaration of function 'main__encode_array_T_Array_int'
tcc: error: undefined symbol 'main__encode_array_T_Array_int'
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

```v
struct IntArray {
	a []int
}

fn encode_array[U](val []U) {
	$if U is int {
		dump('is int')
	} $else $if U is $Struct {
		dump('is struct')
	}
}

fn encode_struct[U](val U) {
	$for field in U.fields {
		if field.is_array {
			value := val.$(field.name)
			encode_array(value)
		}
	}
}

fn main() {
	int_array := IntArray{
		a: [1, 2, 3]
	}
	encode_struct(int_array)
}
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.3.2 101aa4a

### Environment details (OS name and version, etc.)

```bash
OS: linux, Ubuntu 22.10
Processor: 16 cpus, 64bit, little endian, AMD Ryzen 7 5800H with Radeon Graphics
CC version: cc (Ubuntu 12.2.0-3ubuntu1) 12.2.0

getwd: /home/hitalo/Documents/v_bug
vmodules: /home/hitalo/.vmodules
vroot: /home/hitalo/v
vexe: /home/hitalo/v/v
vexe mtime: 2022-12-28 11:43:18
is vroot writable: true
is vmodules writable: true
V full version: V 0.3.2 458e68e.101aa4a

Git version: git version 2.37.2
Git vroot status: weekly.2022.52-9-g101aa4ae (1 commit(s) behind V master)
.git/config present: true
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3
```

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1367,7 +1367,12 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 					&& has_comptime_field {
 					mut concrete_types := node.concrete_types.map(g.unwrap_generic(it))
 					for k in comptime_args {
-						concrete_types[k] = g.comptime_for_field_type
+						arg_sym := g.table.sym(g.comptime_for_field_type)
+						if arg_sym.kind == .array {
+							concrete_types[k] = (arg_sym.info as ast.Array).elem_type
+						} else {
+							concrete_types[k] = g.comptime_for_field_type
+						}
 					}
 					name = g.generic_fn_name(concrete_types, name)
 				} else {

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1366,8 +1366,8 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 				if g.comptime_for_field_type != 0 && g.inside_comptime_for_field
 					&& has_comptime_field {
 					mut concrete_types := node.concrete_types.map(g.unwrap_generic(it))
+					arg_sym := g.table.sym(g.comptime_for_field_type)
 					for k in comptime_args {
-						arg_sym := g.table.sym(g.comptime_for_field_type)
 						if arg_sym.kind == .array {
 							concrete_types[k] = (arg_sym.info as ast.Array).elem_type
 						} else {

--- a/vlib/v/tests/fn_call_comptime_array_arg_test.v
+++ b/vlib/v/tests/fn_call_comptime_array_arg_test.v
@@ -1,0 +1,30 @@
+struct IntArray {
+	a []int
+}
+
+fn encode_array[U](val []U) []string {
+	mut out := []string{}
+	$if U is int {
+		out << 'is int'
+	} $else $if U is $Struct {
+		out << 'is struct'
+	}
+	return out
+}
+
+fn encode_struct[U](val U) ?[]string {
+	$for field in U.fields {
+		if field.is_array {
+			value := val.$(field.name)
+			return encode_array(value)
+		}
+	}
+	return none
+}
+
+fn test_main() {
+	int_array := IntArray{
+		a: [1, 2, 3]
+	}
+	assert encode_struct(int_array)?.str() == "['is int']"
+}

